File: 代码/问题二/code/chapter3_data_quality_figures.py
#!/usr/bin/env python3
"""Draw Chapter 3 data-quality figures for Brent crude data.

Generated figures:
  - Fig. 3-1 missing-value counts by field
  - Fig. 3-2 boxplot of Brent close prices

The script uses only Python standard library modules and writes SVG files.
"""

from __future__ import annotations

import csv
import math
from pathlib import Path


ROOT = Path(__file__).resolve().parents[3]
CSV_PATH = ROOT / "附件1.布伦特原油期货主力合约价格数据.csv"
OUT_DIR = ROOT / "代码" / "问题二" / "results" / "chapter3_figures"
OUT_DIR.mkdir(parents=True, exist_ok=True)

FIELD_LABELS = {
    "time": "交易日期",
    "thscode": "合约代码",
    "preClose": "前收盘价",
    "open": "开盘价",
    "high": "最高价",
    "low": "最低价",
    "close": "收盘价",
}


def is_missing(value: str) -> bool:
    return value is None or value.strip() == "" or value.strip().upper() == "NA"


def read_data() -> list[dict[str, str]]:
    with CSV_PATH.open("r", encoding="utf-8-sig", newline="") as file:
        return list(csv.DictReader(file))


def percentile(values: list[float], p: float) -> float:
    if not values:
        raise ValueError("empty values")
    xs = sorted(values)
    pos = (len(xs) - 1) * p
    lo = math.floor(pos)
    hi = math.ceil(pos)
    if lo == hi:
        return xs[lo]
    return xs[lo] * (hi - pos) + xs[hi] * (pos - lo)


def svg_text(x: float, y: float, text: str, *, size=14, anchor="middle", weight="normal", rotate=None) -> str:
    transform = f' transform="rotate({rotate} {x} {y})"' if rotate is not None else ""
    return (
        f'<text x="{x:.1f}" y="{y:.1f}" text-anchor="{anchor}" dominant-baseline="middle"'
        f'{transform} font-family="Songti SC, SimSun, Microsoft YaHei, PingFang SC, Arial Unicode MS, serif"'
        f' font-size="{size}" font-weight="{weight}">{text}</text>'
    )


def draw_missing_bar(rows: list[dict[str, str]]) -> Path:
    fields = list(rows[0].keys())
    counts = [sum(1 for row in rows if is_missing(row.get(field, ""))) for field in fields]

    width, height = 980, 580
    left, right, top, bottom = 88, 38, 70, 105
    plot_w = width - left - right
    plot_h = height - top - bottom
    max_count = max(max(counts), 1)
    y_max = max_count if max_count <= 5 else math.ceil(max_count / 5) * 5

    def sx(i: int) -> float:
        return left + (i + 0.5) * plot_w / len(fields)

    def sy(v: float) -> float:
        return top + (y_max - v) / y_max * plot_h

    parts = [
        f'<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}">',
        '<rect width="100%" height="100%" fill="white"/>',
        svg_text(width / 2, 32, "图3-1 布伦特原油价格数据缺失值统计图", size=22, weight="bold"),
    ]

    # Grid and axis.
    for tick in range(0, y_max + 1):
        y = sy(tick)
        parts.append(f'<line x1="{left}" y1="{y:.1f}" x2="{width-right}" y2="{y:.1f}" stroke="#E5E5E5"/>')
        parts.append(svg_text(left - 12, y, str(tick), size=12, anchor="end"))
    parts.append(f'<line x1="{left}" y1="{height-bottom}" x2="{width-right}" y2="{height-bottom}" stroke="#333" stroke-width="1.5"/>')
    parts.append(f'<line x1="{left}" y1="{top}" x2="{left}" y2="{height-bottom}" stroke="#333" stroke-width="1.5"/>')

    bar_w = plot_w / len(fields) * 0.58
    for i, (field, count) in enumerate(zip(fields, counts)):
        x = sx(i) - bar_w / 2
        y = sy(count)
        bar_h = height - bottom - y
        parts.append(f'<rect x="{x:.1f}" y="{y:.1f}" width="{bar_w:.1f}" height="{bar_h:.1f}" fill="#4E79A7"/>')
        parts.append(svg_text(sx(i), y - 14, str(count), size=13, weight="bold"))
        parts.append(svg_text(sx(i), height - bottom + 28, FIELD_LABELS.get(field, field), size=13, rotate=-25))

    parts.append(svg_text(width / 2, height - 22, "字段名称", size=15))
    parts.append(svg_text(24, height / 2, "缺失值数量", size=15, rotate=-90))
    parts.append("</svg>")

    path = OUT_DIR / "fig3_1_missing_values.svg"
    path.write_text("\n".join(parts), encoding="utf-8")
    return path


def draw_close_boxplot(rows: list[dict[str, str]]) -> Path:
    closes = [float(row["close"]) for row in rows if not is_missing(row.get("close", ""))]
    q1 = percentile(closes, 0.25)
    q2 = percentile(closes, 0.50)
    q3 = percentile(closes, 0.75)
    iqr = q3 - q1
    lower_fence = q1 - 1.5 * iqr
    upper_fence = q3 + 1.5 * iqr
    whisker_low = min(v for v in closes if v >= lower_fence)
    whisker_high = max(v for v in closes if v <= upper_fence)
    outliers = [v for v in closes if v < lower_fence or v > upper_fence]

    width, height = 720, 620
    left, right, top, bottom = 105, 70, 70, 88
    plot_h = height - top - bottom
    y_min = math.floor(min(closes) / 10) * 10
    y_max = math.ceil(max(closes) / 10) * 10

    def sy(v: float) -> float:
        return top + (y_max - v) / (y_max - y_min) * plot_h

    cx = (width - left - right) / 2 + left
    box_w = 170
    parts = [
        f'<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}">',
        '<rect width="100%" height="100%" fill="white"/>',
        svg_text(width / 2, 32, "图3-2 布伦特原油收盘价箱线图", size=22, weight="bold"),
    ]

    for tick in range(int(y_min), int(y_max) + 1, 10):
        y = sy(tick)
        parts.append(f'<line x1="{left}" y1="{y:.1f}" x2="{width-right}" y2="{y:.1f}" stroke="#E5E5E5"/>')
        parts.append(svg_text(left - 12, y, str(tick), size=12, anchor="end"))
    parts.append(f'<line x1="{left}" y1="{height-bottom}" x2="{width-right}" y2="{height-bottom}" stroke="#333" stroke-width="1.5"/>')
    parts.append(f'<line x1="{left}" y1="{top}" x2="{left}" y2="{height-bottom}" stroke="#333" stroke-width="1.5"/>')

    # Whiskers.
    parts.append(f'<line x1="{cx:.1f}" y1="{sy(whisker_high):.1f}" x2="{cx:.1f}" y2="{sy(q3):.1f}" stroke="#333" stroke-width="2"/>')
    parts.append(f'<line x1="{cx:.1f}" y1="{sy(q1):.1f}" x2="{cx:.1f}" y2="{sy(whisker_low):.1f}" stroke="#333" stroke-width="2"/>')
    parts.append(f'<line x1="{cx-box_w/3:.1f}" y1="{sy(whisker_high):.1f}" x2="{cx+box_w/3:.1f}" y2="{sy(whisker_high):.1f}" stroke="#333" stroke-width="2"/>')
    parts.append(f'<line x1="{cx-box_w/3:.1f}" y1="{sy(whisker_low):.1f}" x2="{cx+box_w/3:.1f}" y2="{sy(whisker_low):.1f}" stroke="#333" stroke-width="2"/>')

    # IQR box and median.
    parts.append(
        f'<rect x="{cx-box_w/2:.1f}" y="{sy(q3):.1f}" width="{box_w:.1f}" height="{sy(q1)-sy(q3):.1f}" '
        'fill="#F4B183" stroke="#333" stroke-width="2"/>'
    )
    parts.append(f'<line x1="{cx-box_w/2:.1f}" y1="{sy(q2):.1f}" x2="{cx+box_w/2:.1f}" y2="{sy(q2):.1f}" stroke="#C00000" stroke-width="3"/>')

    # Outliers with deterministic horizontal jitter.
    for i, value in enumerate(outliers):
        jitter = ((i % 9) - 4) * 5.0
        parts.append(f'<circle cx="{cx+jitter:.1f}" cy="{sy(value):.1f}" r="3.2" fill="white" stroke="#4E79A7" stroke-width="1.5"/>')

    parts.append(svg_text(cx, height - bottom + 32, "收盘价", size=15))
    parts.append(svg_text(28, height / 2, "价格（美元/桶）", size=15, rotate=-90))
    note = f"Q1={q1:.2f}  中位数={q2:.2f}  Q3={q3:.2f}  异常点={len(outliers)}个"
    parts.append(svg_text(width / 2, height - 24, note, size=13))
    parts.append("</svg>")

    path = OUT_DIR / "fig3_2_close_boxplot.svg"
    path.write_text("\n".join(parts), encoding="utf-8")
    return path


def main() -> None:
    rows = read_data()
    fig1 = draw_missing_bar(rows)
    fig2 = draw_close_boxplot(rows)
    print(f"已生成：{fig1}")
    print(f"已生成：{fig2}")


if __name__ == "__main__":
    main()

File: 代码/问题二/code/chapter3_data_quality_figures.m
% Chapter 3 data-quality figures for Brent crude oil price data.
% Generated figures:
%   图3-1 布伦特原油价格数据缺失值统计图
%   图3-2 布伦特原油收盘价箱线图
%   图3-3 布伦特原油交易日期间隔统计图
%   图3-4 标准化后主要价格指标对比图
%   图3-5 布伦特原油收盘价走势分析图
%
% Run:
%   run("/Users/hybuzhy/Documents/Mathematical modeling/代码/问题二/code/chapter3_data_quality_figures.m")

clear; clc;

scriptDir = fileparts(mfilename("fullpath"));
problemDir = fileparts(scriptDir);
rootDir = fileparts(fileparts(problemDir));
csvPath = fullfile(rootDir, "附件1.布伦特原油期货主力合约价格数据.csv");
outDir = fullfile(problemDir, "results", "chapter3_figures_matlab");

if ~exist(outDir, "dir")
    mkdir(outDir);
end

opts = detectImportOptions(csvPath, "Encoding", "UTF-8");
opts = setvaropts(opts, opts.VariableNames, "TreatAsMissing", ["NA", ""]);
T = readtable(csvPath, opts);

% Use Chinese labels for paper figures.
fieldNames = ["交易日期", "合约代码", "前收盘价", "开盘价", "最高价", "最低价", "收盘价"];
rawNames = ["time", "thscode", "preClose", "open", "high", "low", "close"];

if isdatetime(T.time)
    tradeDate = T.time;
else
    tradeDate = datetime(string(T.time), "InputFormat", "yyyy/M/d");
end
[tradeDate, sortIdx] = sort(tradeDate);
T = T(sortIdx, :);

%% 图3-1 布伦特原油价格数据缺失值统计图
missingCounts = zeros(1, numel(rawNames));
for i = 1:numel(rawNames)
    missingCounts(i) = sum(ismissing(T.(rawNames(i))));
end

figure("Position", [100, 100, 980, 560], "Color", "w");
b = bar(missingCounts, 0.58);
b.FaceColor = [0.1216, 0.4667, 0.7059];
grid on;
box on;
set(gca, "XTick", 1:numel(fieldNames), "XTickLabel", fieldNames);
xtickangle(25);
ylabel("缺失值数量");
xlabel("字段名称");
title("图3-1 布伦特原油价格数据缺失值统计图");
ylim([0, max([1, missingCounts]) * 1.25]);

for i = 1:numel(missingCounts)
    text(i, missingCounts(i) + 0.04 * max([1, missingCounts]), ...
        sprintf("%d", missingCounts(i)), ...
        "HorizontalAlignment", "center", "FontSize", 11);
end

exportgraphics(gcf, fullfile(outDir, "fig3_1_missing_values.png"), "Resolution", 300);
exportgraphics(gcf, fullfile(outDir, "fig3_1_missing_values.pdf"), "ContentType", "vector");

%% 图3-2 布伦特原油收盘价箱线图
closePrice = T.close;
closePrice = closePrice(~ismissing(closePrice));

q1 = localPercentile(closePrice, 25);
q2 = localPercentile(closePrice, 50);
q3 = localPercentile(closePrice, 75);
iqrValue = q3 - q1;
lowerFence = q1 - 1.5 * iqrValue;
upperFence = q3 + 1.5 * iqrValue;
whiskerLow = min(closePrice(closePrice >= lowerFence));
whiskerHigh = max(closePrice(closePrice <= upperFence));
outlierMask = closePrice < lowerFence | closePrice > upperFence;
outlierValues = closePrice(outlierMask);
outlierCount = numel(outlierValues);

figure("Position", [100, 100, 760, 580], "Color", "w");
hold on;
grid on;
box on;

% Manually draw a boxplot using only base MATLAB graphics.
boxX = 1;
boxWidth = 0.34;
capWidth = 0.24;
rectangle("Position", [boxX - boxWidth / 2, q1, boxWidth, q3 - q1], ...
    "FaceColor", [0.9569, 0.6980, 0.5137], ...
    "EdgeColor", [0.2, 0.2, 0.2], ...
    "LineWidth", 1.6);
plot([boxX - boxWidth / 2, boxX + boxWidth / 2], [q2, q2], ...
    "Color", [0.75, 0.0, 0.0], "LineWidth", 2.2);
plot([boxX, boxX], [q3, whiskerHigh], "Color", [0.2, 0.2, 0.2], "LineWidth", 1.5);
plot([boxX, boxX], [q1, whiskerLow], "Color", [0.2, 0.2, 0.2], "LineWidth", 1.5);
plot([boxX - capWidth / 2, boxX + capWidth / 2], [whiskerHigh, whiskerHigh], ...
    "Color", [0.2, 0.2, 0.2], "LineWidth", 1.5);
plot([boxX - capWidth / 2, boxX + capWidth / 2], [whiskerLow, whiskerLow], ...
    "Color", [0.2, 0.2, 0.2], "LineWidth", 1.5);

if outlierCount > 0
    jitter = mod((1:outlierCount)' - 1, 9) - 4;
    jitter = jitter * 0.012;
    scatter(boxX + jitter, outlierValues, 18, ...
        "MarkerEdgeColor", [0.1216, 0.4667, 0.7059], ...
        "MarkerFaceColor", "w", ...
        "LineWidth", 1.0);
end

xlim([0.45, 1.55]);
ylim([min(closePrice) - 0.10 * (max(closePrice) - min(closePrice)), ...
    max(closePrice) + 0.10 * (max(closePrice) - min(closePrice))]);
set(gca, "XTick", 1, "XTickLabel", {'收盘价'});
ylabel("价格（美元/桶）");
title("图3-2 布伦特原油收盘价箱线图");

annotationText = sprintf("Q1=%.2f, 中位数=%.2f, Q3=%.2f, 异常点=%d个", ...
    q1, q2, q3, outlierCount);
text(1, min(closePrice) - 0.06 * (max(closePrice) - min(closePrice)), annotationText, ...
    "HorizontalAlignment", "center", "FontSize", 10);
hold off;

exportgraphics(gcf, fullfile(outDir, "fig3_2_close_boxplot.png"), "Resolution", 300);
exportgraphics(gcf, fullfile(outDir, "fig3_2_close_boxplot.pdf"), "ContentType", "vector");

%% 图3-3 布伦特原油交易日期间隔统计图
dateIntervals = days(diff(tradeDate));
intervalCounts = [
    sum(dateIntervals == 1)
    sum(dateIntervals == 2)
    sum(dateIntervals == 3)
    sum(dateIntervals >= 4)
];
intervalLabels = ["1天", "2天", "3天", "4天及以上"];

figure("Position", [100, 100, 860, 520], "Color", "w");
b = bar(intervalCounts, 0.58);
b.FaceColor = [0.4667, 0.6745, 0.1882];
grid on;
box on;
set(gca, "XTick", 1:numel(intervalLabels), "XTickLabel", intervalLabels);
xlabel("相邻交易记录日期间隔");
ylabel("出现次数");
title("图3-3 布伦特原油交易日期间隔统计图");
ylim([0, max(intervalCounts) * 1.18]);
for i = 1:numel(intervalCounts)
    text(i, intervalCounts(i) + 0.03 * max(intervalCounts), ...
        sprintf("%d", intervalCounts(i)), ...
        "HorizontalAlignment", "center", "FontSize", 11);
end
exportgraphics(gcf, fullfile(outDir, "fig3_3_date_interval_counts.png"), "Resolution", 300);
exportgraphics(gcf, fullfile(outDir, "fig3_3_date_interval_counts.pdf"), "ContentType", "vector");

%% 图3-4 标准化后主要价格指标对比图
priceNames = ["open", "high", "low", "close"];
priceLabels = ["开盘价", "最高价", "最低价", "收盘价"];
Z = zeros(height(T), numel(priceNames));
for i = 1:numel(priceNames)
    x = T.(priceNames(i));
    Z(:, i) = (x - min(x)) ./ (max(x) - min(x));
end

figure("Position", [100, 100, 980, 560], "Color", "w");
plot(tradeDate, Z, "LineWidth", 1.1);
grid on;
box on;
xlabel("交易日期");
ylabel("Min-Max标准化值");
title("图3-4 标准化后主要价格指标对比图");
legend(priceLabels, "Location", "northwest");
exportgraphics(gcf, fullfile(outDir, "fig3_4_normalized_price_indicators.png"), "Resolution", 300);
exportgraphics(gcf, fullfile(outDir, "fig3_4_normalized_price_indicators.pdf"), "ContentType", "vector");

%% 图3-5 布伦特原油收盘价走势分析图
figure("Position", [100, 100, 980, 560], "Color", "w");
plot(tradeDate, T.close, "LineWidth", 1.3, "Color", [0.1216, 0.4667, 0.7059]);
grid on;
box on;
xlabel("交易日期");
ylabel("收盘价（美元/桶）");
title("图3-5 布伦特原油收盘价走势分析图");

[minClose, minIdx] = min(T.close);
[maxClose, maxIdx] = max(T.close);
hold on;
scatter(tradeDate(minIdx), minClose, 50, [0.4667, 0.6745, 0.1882], "filled");
scatter(tradeDate(maxIdx), maxClose, 50, [0.8392, 0.1529, 0.1569], "filled");
text(tradeDate(minIdx), minClose, sprintf("  最低 %.2f", minClose), ...
    "VerticalAlignment", "top", "FontSize", 10);
text(tradeDate(maxIdx), maxClose, sprintf("  最高 %.2f", maxClose), ...
    "VerticalAlignment", "bottom", "FontSize", 10);
hold off;
exportgraphics(gcf, fullfile(outDir, "fig3_5_close_price_trend.png"), "Resolution", 300);
exportgraphics(gcf, fullfile(outDir, "fig3_5_close_price_trend.pdf"), "ContentType", "vector");

fprintf("缺失值统计：\n");
for i = 1:numel(rawNames)
    fprintf("  %s: %d\n", fieldNames(i), missingCounts(i));
end
fprintf("收盘价箱线图统计：Q1=%.2f, Median=%.2f, Q3=%.2f, Outliers=%d\n", ...
    q1, q2, q3, outlierCount);
fprintf("交易日期范围：%s 至 %s，共 %d 条记录\n", ...
    datestr(tradeDate(1), "yyyy-mm-dd"), datestr(tradeDate(end), "yyyy-mm-dd"), height(T));
fprintf("MATLAB figures written to:\n%s\n", outDir);

function value = localPercentile(x, p)
    x = sort(x(:));
    n = numel(x);
    if n == 0
        error("Cannot compute percentile of an empty vector.");
    end
    pos = 1 + (n - 1) * p / 100;
    lo = floor(pos);
    hi = ceil(pos);
    if lo == hi
        value = x(lo);
    else
        value = x(lo) * (hi - pos) + x(hi) * (pos - lo);
    end
end
